### PR TITLE
Add Any type and to_any/from_any primitives to protobuf

### PR DESCRIPTION
**Pull request by cocreature**
Following a suggestion by Rémy, the protobuf representation is more
general and is associated with an arbitrary type instead of a
typecon. This allows us to easily extend this later to a full Any
type.

I’ve still called the type in the protobuf Any instead of Haskell’s
Dynamic since I find AnyTemplate more clear than DynamicTemplate and
having AnyTemplate and Dynamic seems confusing.

Right now, the decoder enforces that the type is a TypeCon.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -1467,6 +1485,11 @@ need to be evaluated further. ::
    ——————————————————————————————————————————————————— ValExpTupleCon
      ⊢ᵥ  ⟨ f₁ = e₁, …, fₘ = eₘ ⟩
 
+
+     ⊢ᵥ e
+   ——————————————————————————————————————————————————— ValExpToAnyTemplate
+     ⊢ᵥ  'to_any_template' e
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
     ⊢ᵥ  'to_any_template' @Mod:T e
```

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -1625,6 +1648,15 @@ exact output.
     —————————————————————————————————————————————————————————————————————— EvExpLet
       'let' x : τ = e₁ 'in' e₂ ‖ E₀  ⇓  r ‖ E₂
 
+      e₁ ‖ E₀  ⇓  Ok (to_any_template @Mod:T v₁) ‖ E₁
+    —————————————————————————————————————————————————————————————————————— EvExpFromAnyTemplateSucc
+      'from_any_template' @Mod:T e₁ ‖ E₀  ⇓  'Some' @Mod:T v₁ ‖ E₁
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      'from_any_template' @Mod:T e ‖ E₀  ⇓  'Some' @Mod:T v ‖ E₁
```

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -1625,6 +1648,15 @@ exact output.
...
+
+      e₁ ‖ E₀  ⇓  Ok (to_any_template @Mod₂:T₂ v₂) ‖ E₁, Mod₁:T₁ ≠ Mod₂:T₂
+    —————————————————————————————————————————————————————————————————————— EvExpFromAnyTemplateFail
+      'from_any_template' @Mod₁:T₁ e₁ ‖ E₀  ⇓  'None' ‖ E₁
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      'from_any_template' @Mod₁:T₁ e ‖ E₀  ⇓  'None' ‖ E₁
```

File: daml-lf/spec/daml-lf-1.rst [hunk 4]
```diff
@@ -1625,6 +1648,15 @@ exact output.
     —————————————————————————————————————————————————————————————————————— EvExpLet
       'let' x : τ = e₁ 'in' e₂ ‖ E₀  ⇓  r ‖ E₂
 
+      e₁ ‖ E₀  ⇓  Ok (to_any_template @Mod:T v₁) ‖ E₁
+    —————————————————————————————————————————————————————————————————————— EvExpFromAnyTemplateSucc
+      'from_any_template' @Mod:T e₁ ‖ E₀  ⇓  'Some' @Mod:T v₁ ‖ E₁
+
+      e₁ ‖ E₀  ⇓  Ok (to_any_template @Mod₂:T₂ v₂) ‖ E₁, Mod₁:T₁ ≠ Mod₂:T₂
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      e ‖ E₀  ⇓  Ok ('to_any_template' @Mod₂:T₂ v) ‖ E₁     Mod₁:T₁ ≠ Mod₂:T₂
```

File: daml-lf/archive/da/daml_lf_1.proto [hunk 5]
```diff
@@ -234,6 +234,9 @@ enum PrimType {
   // *Available in versions >= 1.dev*
   NUMERIC = 17;
 
+  // Builtin type 'Any'
+  // *Available in versions >= 1.dev*
+  ANY = 18;
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/da/daml_lf_1.proto]: Scala people would prefer not using `Any`. (cc @gerolf-da)

File: daml-lf/spec/daml-lf-1.rst [hunk 6]
```diff
@@ -1625,6 +1648,15 @@ exact output.
     —————————————————————————————————————————————————————————————————————— EvExpLet
       'let' x : τ = e₁ 'in' e₂ ‖ E₀  ⇓  r ‖ E₂
 
+      e₁ ‖ E₀  ⇓  Ok (to_any_template @Mod:T v₁) ‖ E₁
```
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
      e ‖ E₀  ⇓  Ok ('to_any_template' @Mod:T v) ‖ E₁
```

Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst]: We also need the rule
```
      e ‖ E₀  ⇓  Ok v ‖ E₁    —————————————————————————————————————————————————————————————————————— EvExpLet
      'to_any_template' @Mod:T e ‖ E₀  ⇓  Ok('to_any_template' @Mod:T v) ‖ E₁    ```

File: daml-lf/archive/da/daml_lf_1.proto (hunk 5 above)
Comment (cocreature) [reply daml-lf/archive/da/daml_lf_1.proto]: So what are you proposing instead? Calling this `Dynamic` while keeping `AnyTemplate` seems really confusing and I’m not entirely sold that `DynamicTemplate` is a good name.

Comment (cocreature) [reply daml-lf/archive/da/daml_lf_1.proto]: I’ll go for `Dynamic` now and keep `AnyTemplate`. I’ll rename it in a separate PR to not clutter the diff here too much.

Comment (remyhaemmerle-da) [reply daml-lf/archive/da/daml_lf_1.proto]: OK let's go with `ANY`.

**Comment by gerolf-da** [comment]
Thanks, @cocreature!

**Pull request merged**